### PR TITLE
Fix callback issues

### PR DESCRIPTION
**Pull request by dh031200**
- [x] 1. add test for `add_callback`
- [x] 2. fix callbacks by reference 

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Implemented callback functionality for model training and inference in the Ultralytics YOLO framework.

### 📊 Key Changes
- Introduced a `test_export()` function for testing the new export callback mechanism.
- Added test functions to check callback functionality in `test_detect()`, `test_segment()`, and `test_classify()`.
- Removed `defaultdict` use from several modules (exporter, predictor, trainer, validator) and replaced it with a function `get_default_callbacks()`.
- Added callback parameter `_callbacks` to constructors to allow passing callbacks during instantiation.
- Defined `add_callback()` method in multiple classes to append new callbacks to the events list.

### 🎯 Purpose & Impact
- **Enhances Flexibility**: The newly introduced callbacks allow users to add custom behavior at various points during model training, inference, and exporting processes.
- **Improves Code Quality**: By replacing `defaultdict` with a function, the code becomes cleaner and more manageable.
- **Augments Customization**: Users can now receive specific events and react with custom functions, making the framework more intuitive and adaptable to different workflows.

### CODE REVIEW

**Comment by dh031200** [comment]
@glenn-jocher @AyushExel 
Originally, only `trainer` has add_callback function. So, if I put test code in `validator` and `predictor`, CI/Test will fail.

**Comment by dh031200** [comment]
All tests passed, so I will also add validator and predictor tests.

**Comment by glenn-jocher** [comment]
@dh031200 it sounds like you are proposing to add tests for the `add_callback` function in the `validator` and `predictor` classes. However, you are concerned that doing so may cause issues because the `add_callback` function originally only existed in the `trainer` class. Please clarify if you would like guidance on how to properly add tests for these classes or if you no longer need assistance.

**Comment by glenn-jocher** [comment]
@dh031200 hey the tests look great, this is exactly what we need I think. 

Regarding the actual callback modifications, I'll leave this to @AyushExel to review since he has more experience with our callbacks.

**Comment by dh031200** [comment]
@glenn-jocher `Predictor` needs `add_callback` function because of #1781 
`Trainer` and `Predictor` have `add_callbacks` function but only `Validator` does not have function, so it seems inconsistent.
So I added `add_callback`function to the `Validator` to give consistency and flexibility to users who use the `engine` directly like #1845.

**Comment by glenn-jocher** [comment]
@dh031200 it makes sense to add `add_callback` to the `Validator` class to maintain consistency across all classes providing this functionality to users. It is also helpful for users of the `engine` directly, which can now use `Validator` in the same way as they use `Trainer` and `Predictor`. This should increase flexibility for users and make the use of the engine more intuitive.

**Comment by AyushExel** [comment]
@dh031200 I think it looks good and the tests are passing. Just to confirm, this does pretty much the same thing as the previous PR but now base engine executors accept a  callbacks instead of using the global callbacks right?

**Comment by dh031200** [comment]
@AyushExel Yes, that's right. So they use their own callbacks, not global callbacks.

**Comment by AyushExel** [comment]
Okay perfect! Looks good then

**Comment by maltelorbach** [comment]
@AyushExel @glenn-jocher May I suggest to add the same to the Exporter class(es) as well. These also have callbacks that would now not be called when calling `model.export()`.

**Comment by dh031200** [comment]
@maltelorbach Okay, I added _callbacks argument into exporter. Also added test for exporter. 😊

**Comment by glenn-jocher** [comment]
@dh031200, that's great to hear that you have added `_callbacks` argument to the `exporter` class and have added a test for it as well. This will enable `Exporter` class to use its own callbacks like `Trainer`, `Validator` and `Predictor`. This also ensures that users can specify specific callbacks for `Exporter` class without affecting global callbacks. Thank you for your contribution!

File: ultralytics/yolo/engine/model.py [hunk 1]
```diff
@@ -78,7 +79,7 @@ def __init__(self, model: Union[str, Path] = 'yolov8n.pt', task=None) -> None:
             task (Any, optional): Task type for the YOLO model. Defaults to None.
 
         """
-        self._reset_callbacks()
+        self.callbacks = deepcopy(callbacks.default_callbacks)
```
Comment (glenn-jocher) [inline ultralytics/yolo/engine/model.py:82]: @dh031200 do we require deepcopy() here? Can you we simply use the dictionary built-in .copy() method, i.e. here is my proposed change:
```diff
- from copy import deepcopy
- self.callbacks = deepcopy(callbacks.default_callbacks)
+ self.callbacks = callbacks.default_callbacks.copy()
```

**Comment by dh031200** [comment]
@glenn-jocher I tested `.copy()` and `deepcopy()`. they show different results.
And I worked on the `base.py` to prevent duplicate code.

Testcode
-----
```python
from tests.test_engine import *

test_export()
test_detect()
test_segment()
test_classify()
```

Results
--------
copy() :
```
YOLOv8n-cls summary: 99 layers, 1451098 parameters, 1451098 gradients, 3.4 GFLOPs
optimizer: SGD(lr=0.01) with parameter groups 26 weight(decay=0.0), 27 weight(decay=0.0005), 27 bias
callback test passed
callback test passed
callback test passed
Image sizes 32 train, 32 val
Using 0 dataloader workers
Logging results to /mnt/d/Project/pr/revert/ultralytics/runs/classify/train11
Starting training for 1 epochs...

      Epoch    GPU_mem       loss  Instances       Size
        1/1         0G     0.3573         10         32: 100%|██████████| 1/1 [00:00<00:00,  5.69it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 1/1 [00:00<00:00,  6.54it/s]
                   all        0.1        0.5

1 epochs completed in 0.000 hours.
Optimizer stripped from /mnt/d/Project/pr/revert/ultralytics/runs/classify/train11/weights/last.pt, 3.0MB
Optimizer stripped from /mnt/d/Project/pr/revert/ultralytics/runs/classify/train11/weights/best.pt, 3.0MB
Results saved to /mnt/d/Project/pr/revert/ultralytics/runs/classify/train11
callback test passed
callback test passed
callback test passed
Ultralytics YOLOv8.0.67 🚀 Python-3.10.10 torch-2.0.0+cu117 CPU
YOLOv8n-cls summary (fused): 73 layers, 1447690 parameters, 0 gradients, 3.3 GFLOPs
               classes   top1_acc   top5_acc: 100%|██████████| 1/1 [00:00<00:00,  6.11it/s]
                   all        0.1        0.5
Speed: 0.0ms preprocess, 0.7ms inference, 0.0ms loss, 0.0ms postprocess per image
Results saved to /mnt/d/Project/pr/revert/ultralytics/runs/segment/train30

Ultralytics YOLOv8.0.67 🚀 Python-3.10.10 torch-2.0.0+cu117 CPU
YOLOv8n-cls summary (fused): 73 layers, 1447690 parameters, 0 gradients, 3.3 GFLOPs
callback test passed
callback test passed
callback test passed
image 1/2 /mnt/d/Dev/ultralytics/ultralytics/assets/bus.jpg: 64x64 stingray 0.10, ostrich 0.10, hen 0.10, goldfish 0.10, hammerhead 0.10, 2.4ms
image 2/2 /mnt/d/Dev/ultralytics/ultralytics/assets/zidane.jpg: 64x64 stingray 0.10, ostrich 0.10, hen 0.10, goldfish 0.10, hammerhead 0.10, 3.5ms
Speed: 0.1ms preprocess, 2.9ms inference, 0.0ms postprocess per image at shape (1, 3, 64, 64)
Results saved to /mnt/d/Project/pr/revert/ultralytics/runs/detect/train85
```

deepcopy():
```
YOLOv8n-cls summary: 99 layers, 1451098 parameters, 1451098 gradients, 3.4 GFLOPs
optimizer: SGD(lr=0.01) with parameter groups 26 weight(decay=0.0), 27 weight(decay=0.0005), 27 bias
callback test passed
Image sizes 32 train, 32 val
Using 0 dataloader workers
Logging results to /mnt/d/Project/pr/revert/ultralytics/runs/classify/train13
Starting training for 1 epochs...

      Epoch    GPU_mem       loss  Instances       Size
        1/1         0G     0.3573         10         32: 100%|██████████| 1/1 [00:00<00:00,  5.52it/s]
               classes   top1_acc   top5_acc: 100%|██████████| 1/1 [00:00<00:00,  6.43it/s]
                   all        0.1        0.5

1 epochs completed in 0.000 hours.
Optimizer stripped from /mnt/d/Project/pr/revert/ultralytics/runs/classify/train13/weights/last.pt, 3.0MB
Optimizer stripped from /mnt/d/Project/pr/revert/ultralytics/runs/classify/train13/weights/best.pt, 3.0MB
Results saved to /mnt/d/Project/pr/revert/ultralytics/runs/classify/train13
callback test passed
Ultralytics YOLOv8.0.67 🚀 Python-3.10.10 torch-2.0.0+cu117 CPU
YOLOv8n-cls summary (fused): 73 layers, 1447690 parameters, 0 gradients, 3.3 GFLOPs
               classes   top1_acc   top5_acc: 100%|██████████| 1/1 [00:00<00:00,  6.10it/s]
                   all        0.1        0.5
Speed: 0.0ms preprocess, 0.7ms inference, 0.0ms loss, 0.0ms postprocess per image
Results saved to /mnt/d/Project/pr/revert/ultralytics/runs/segment/train36

Ultralytics YOLOv8.0.67 🚀 Python-3.10.10 torch-2.0.0+cu117 CPU
YOLOv8n-cls summary (fused): 73 layers, 1447690 parameters, 0 gradients, 3.3 GFLOPs
callback test passed
image 1/2 /mnt/d/Dev/ultralytics/ultralytics/assets/bus.jpg: 64x64 stingray 0.10, ostrich 0.10, hen 0.10, goldfish 0.10, hammerhead 0.10, 3.6ms
image 2/2 /mnt/d/Dev/ultralytics/ultralytics/assets/zidane.jpg: 64x64 stingray 0.10, ostrich 0.10, hen 0.10, goldfish 0.10, hammerhead 0.10, 2.4ms
Speed: 0.1ms preprocess, 3.0ms inference, 0.0ms postprocess per image at shape (1, 3, 64, 64)
Results saved to /mnt/d/Project/pr/revert/ultralytics/runs/detect/train99
```

**Comment by glenn-jocher** [comment]
@dh031200 thanks for the update! Please revert my deepcopy removal then (apologies!) and this is good to merge :)

**Comment by dh031200** [comment]
@glenn-jocher You're welcome. I think everything is fine. But the CI tests failed with `FileNotFoundError: val: Error loading data from /home/runner/work/ultralytics/datasets/coco8/images/val`. How do I rerun CI tests?

**Comment by glenn-jocher** [comment]
@dh031200 don't worry, these are unrelated issues in `updates` branch that are no resolved. I've restarted CI and believe they should be all green now. 

Is this PR good to merge then?

**Comment by dh031200** [comment]
@glenn-jocher Yes. I think now it's ready to merge.

**Comment by glenn-jocher** [comment]
@dh031200 yayyyy PR merged! 

Thank you @dh031200 and @maltelorbach for your contributions. This change should be live in 8.0.67 shortly.

**Comment by glenn-jocher** [comment]
@dh031200 yayyyy PR merged! 

Thank you @dh031200 and @maltelorbach for your contributions. This change should be live in 8.0.67 shortly.

**Pull request merged**